src/tools/miri/src/shims/foreign_items.rs

```diff
@@ -6,23 +6,32 @@ use rustc_hir::{
     def::DefKind,
     def_id::{CrateNum, LOCAL_CRATE},
 };
-use rustc_middle::middle::{
-    codegen_fn_attrs::CodegenFnAttrFlags, dependency_format::Linkage,
-    exported_symbols::ExportedSymbol,
-};
 use rustc_middle::mir;
 use rustc_middle::ty;
+use rustc_middle::{
+    middle::{
+        codegen_fn_attrs::CodegenFnAttrFlags, dependency_format::Linkage,
+        exported_symbols::ExportedSymbol,
+    },
+    ty::Ty,
+};
 use rustc_session::config::CrateType;
 use rustc_span::Symbol;
 use rustc_target::{
-    abi::{Align, Size},
+    abi::{call::FnAbi, Align, Size},
     spec::abi::Abi,
 };
 
 use super::backtrace::EvalContextExt as _;
 use crate::*;
 use helpers::{ToHost, ToSoft};
 
+#[derive(Debug, Copy, Clone)]
+pub enum ExtraFnVal {
+    ForeignFn { link_name: Symbol },
+    DynSym(DynSym),
+}
```

Comment:
Both of these represent function pointers to a named global symbol. Why do we need to distinguish these variants?

Reply:
I don't know. They were handled differently before and I just kept the same difference — `DynSym` has an assert that makes sure that we actually used a shim, and not an exported symbol.

Reply:
Before, this type had a clear meaning: it represented "things a function pointer can point to". The interpreter natively supports function pointers that point to an `Instance`, but for `dlsym` support we needed function pointers that point to something else, and that's what this extra-fn stuff is about.

But with your change, now there can be a function pointer that points to `DynSym("functionname")`, and a function pointer that points to `ForeignFn { link_name: "functionname" }`, and you made those be different things. That's not pre-existing, that's new in this PR.

Using ExtraFnVal as return type in `find_mir_or_extra_fn` makes some sense, because it matches the same dichotomy: either a native Rust function (`Instance`/MIR body), or "something else". Probably things could be cleaned up a bit by actually using `FnVal<ExtraFnVal>` as return type, then the function pointer ("indirect call"/`Ty::FnPtr`) codepath and the direct call (`Ty::FnDef`) code path can be even more unified.

But that is something we can do later, currently there are bigger problems.

Reply:
> That's not pre-existing

I'm pretty sure it is. `DynSym`s are handled in `emulate_dyn_sym` here: 
https://github.com/rust-lang/rust/blob/0395fa387ad4de63ea1b36703ca0d65017b8c2dc/src/tools/miri/src/shims/foreign_items.rs#L156-L157

While usual foreign function are handled in a branch in `find_mir_or_eval_fn` here:
https://github.com/rust-lang/rust/blob/0395fa387ad4de63ea1b36703ca0d65017b8c2dc/src/tools/miri/src/machine.rs#L960-L960

Those are slightly distinct (see the assert) even though they result in basically the same `emulate_foreign_item` call.

I have nothing against having only a single type, but I'm not sure how to keep this assert then...

Reply:
> I'm pretty sure it is. DynSyms are handled in emulate_dyn_sym here:

> While usual foreign function are handled in a branch in find_mir_or_eval_fn here:

Yes, but that's not what I was talking about. We just had two separate codepaths for handling calls to Rust items vs dynamically loaded items. The "obvious" think for you to do would be having two different types to represent this. You decided to unify these types, which I can get behind, but it's a bad idea to unify them by introducing redundancy *in the value type of the interpreter*.

The assert is a leftover from old code where we had even bigger differences between direct and indirect calls. I'm fine with getting rid of it.